### PR TITLE
feat: SpectralConv2d, SpectralConvTranspose2d

### PR DESCRIPTION
**Pull request by YodaEmbedding**
Introduced in ["Efficient Nonlinear Transforms for Lossy Image Compression"][Balle2018efficient] by Johannes Ballé, PCS 2018. Reparameterizes the weights to be derived from weights stored in the frequency domain. In the original paper, this is referred to as "spectral Adam" or "Sadam" due to its effect on the Adam optimizer update rule. The motivation behind representing the weights in the frequency domain is that optimizer updates/steps may now affect all frequencies to an equal amount. This improves the gradient conditioning, thus leading to faster convergence and increased stability at larger learning rates.

spectral_adam

spectral_adam_rd_curves

For comparison, see the TensorFlow Compression implementations of [`SignalConv2D`] and [`RDFTParameter`]. They seem to use `SignalConv2d` in most of their provided architectures:
https://github.com/search?q=repo%3Atensorflow%2Fcompression+Conv2D&type=code

Furthermore, since this is a simple invertible transformation on the weights, it is trivial to convert any existing pretrained weights into this form via:

```python
weight_transformed = self._to_transform_domain(weight)
```

To override `self.weight` as a property, I'm unregistering the module using `del self._parameters["weight"]` as shown in https://github.com/pytorch/pytorch/issues/46886, and also [using the fact][property-descriptor-so] that `@property` [returns a descriptor object][property-descriptor-docs] so that `self.weight` "falls back" to the property.

```python
    def __init__(self, ...):
        self.weight_transformed = nn.Parameter(self._to_transform_domain(self.weight))
        del self._parameters["weight"]  # Unregister weight, and fallback to property.

    @property
    def weight(self) -> Tensor:
        return self._from_transform_domain(self.weight_transformed)
```

---

Checklist:

 - [ ] Verify: See if training from scratch actually improves performance or training times by appropriately replacing conv/deconv in existing models.
 - [ ] Verify: Check that `SpectralConvTransposed2d` is defined correctly... Since `nn.ConvTransposed2d` inherits from the same `_ConvNd` as `nn.Conv2d` with minor adjustments, I assumed that the weights are treated in the same way.
 - [ ] Ergonomics: If we choose to offer reparametrized versions of already created models, they should override `load_state_dict` to be able to load/write checkpoints in a way that is compatible with the original (non-reparametrized) model definitions.
 - [ ] Performance (medium): See if PyTorch 2.0 Dynamo's [model compilation](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html) works when we manually use 2x `torch.float32` instead of `torch.complex64`. Or maybe it's just the FFT which ruins things. It would be nice if there were a way to skip compilation for certain incompatible operations, particularly since we're merely taking a FFT of leaf tensors, not intermediate computations...
 - [ ] Performance (minor): During inference time, the convolution weights can be precomputed, instead of having to run a FFT on them repeatedly. Perhaps this can be handled inside `CompressionModel.update()`. Or even easier: when the `SpectralConv2d` module is in [`eval` mode], just cache the computed weight, and invalidate/clear the cache when the module switches back to `train` mode.
 - [x] Naming: There's not too many collisions with existing usages of [`SpectralConv2d` on Google](https://www.google.com/search?q=SpectralConv2d), so this should be an OK name. Tensorflow Compression calls their conv layers `SignalConv2d`. Perhaps the most "accurate" name would be `SpectralReparametrizedConv2d`, but that's a bit wordy.

[Balle2018efficient]: https://arxiv.org/abs/1802.00847
[`SignalConv2D`]: https://github.com/tensorflow/compression/blob/v2.14.0/tensorflow_compression/python/layers/signal_conv.py#L61
[`RDFTParameter`]: https://github.com/tensorflow/compression/blob/v2.14.0/tensorflow_compression/python/layers/parameters.py#L71
[property-descriptor-docs]: https://docs.python.org/3/howto/descriptor.html#properties
[property-descriptor-so]: https://stackoverflow.com/a/17330273/365102
[`eval` mode]: https://stackoverflow.com/a/51433411/365102

### CODE REVIEW

**Comment by YodaEmbedding** [comment]
### SpectralConv2d vs Conv2d experiments

Below are a couple of example runs (with different randomly initialized kernels) to compare SpectralConv2d vs Conv2d. (Ignore the inaccurate titles.)

---

sinusoidal channel averaging:

```python
conv_kwargs = dict(in_channels=3, out_channels=2, kernel_size=11, padding=5)

def init_conv_target(conv):
    k = 0.05 * torch.linspace(0, 2 * torch.pi, conv.kernel_size[0], device=device).sin()
    conv.weight.data[:] = k
```

fit_spectralconv2d

fit_spectralconv2d

fit_spectralconv2d

fit_spectralconv2d

---

Simple depthwise cosine "smoothing" kernel:

```python
conv_kwargs = dict(in_channels=8, out_channels=8, kernel_size=11, padding=5)

def init_conv_target(conv):
    k = torch.linspace(0, torch.pi, conv.kernel_size[0], device=device).sin()
    k = k * k[:, None]
    k = k / k.sum()
    idx = torch.arange(conv.in_channels, device=device)
    conv.weight.data[:] = 0
    conv.weight.data[idx, idx, :, :] = k
```

fit_spectralconv2d

---

Simple depthwise vertical edge-detector kernel:

```python
conv_kwargs = dict(in_channels=4, out_channels=4, kernel_size=5, padding=5, groups=4)

def init_conv_target(conv):
    K = conv.kernel_size[0]
    k = 2 * torch.linspace(-1, 1, K, device=device) / K**2
    conv.weight.data[:] = k
```

fit_spectralconv2d

---

Random initialization:

```python
conv_kwargs = dict(in_channels=4, out_channels=4, kernel_size=5, padding=5)

def init_conv_target(conv):
    pass  # Maintain random initialization.
```

fit_spectralconv2d

Surprisingly, SpectralConv2d is not worse...!

---



Figures generated via:

```python
import torch
import torch.nn as nn
import matplotlib.pyplot as plt

from compressai.layers import SpectralConv2d

device = "cuda"


def train(conv, conv_target, max_steps=1000, lr=1e-4, batch_size=16):
    losses = []
    for step in range(max_steps):
        conv.zero_grad()
        x = torch.rand((batch_size, conv.in_channels, 256, 256)).to(device)
        y = conv_target(x).detach()
        y_hat = conv(x)
        loss = ((y - y_hat) ** 2).mean()
        loss.backward()
        for param in conv.parameters():
            param.data -= param.grad * lr
        print(f"step {step:04d}  loss {loss.item():.4f}")
        losses.append(loss.item())
    return losses


models = {
    "spectral": SpectralConv2d(**conv_kwargs),
    "regular": nn.Conv2d(**conv_kwargs),
}

# Initialize all models to exact same random initialization.
conv_rand = nn.Conv2d(**conv_kwargs).to(device)
init_weight = conv_rand.weight.data
init_bias = conv_rand.bias.data

for key, conv in models.items():
    conv.to(device)
    if isinstance(conv, SpectralConv2d):
        conv.weight_transformed.data = conv._to_transform_domain(init_weight).clone()
    else:
        conv.weight.data = init_weight.clone()
    conv.bias.data = init_bias.clone()

conv_target = nn.Conv2d(**conv_kwargs).to(device)
init_conv_target(conv_target)

results = {key: train(conv, conv_target) for key, conv in models.items()}
print("Final losses:")
print("\n".join(f"{key}: {val[-1]:.4f}" for key, val in results.items()))

fig, ax = plt.subplots()
for key, y in results.items():
    ax.plot(y, label=key)
ax.legend()
ax.set(xlabel="step", ylabel="loss", title="Fitting to a target kernel")
fig.savefig("fit_spectralconv2d.png")
```

**Comment by YodaEmbedding** [comment]
### SpectralConv2d vs Conv2d experiments

`++model.name="bmshj2018-factorized" ++criterion.lmbda=0.0067`

The inference/test loss seems to show much quicker convergence with SpectralConv2d:

<table>
<thead>
<tr>
<th>Inference/test loss curves comparison</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
</tr>
<tr>
<td>Blue: Conv2d. Green: SpectralConv2d.</td>
</tr>
</tbody>
</table>

However, even though the *inference/test* loss is evidently much smaller, the position of the point on the RD plot is actually less "optimal" w.r.t. the default configuration:

<table>
<thead>
<tr>
<th>Conv2d</th>
<th>SpectralConv2d</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
</tr>
</tbody>
</table>

The loss function $L = R + \lambda D$ is perhaps not the best way to express the optimal tradeoff. A point with lower $L$ actually has worse (R,D) relative to the optimal achievable (R, D) for a given model architecture?!

One possibility to mitigate this (other than devising a "better" loss function) might be to initially train with the spectral frequency equalization enabled for the first 20 epochs, and then disable it for the remainder of the training.

**Comment by fracape** [comment]
merged together with #270
